### PR TITLE
Refactor logout to use buildLogoutUrl

### PR DESCRIPTION
**Pull request by rnwolfe**
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

This PR seeks to build a logout URL builder function (`buildLogoutUrl`) that the `logout` function then uses. This is in line with the changes made for `buildAuthorizeUrl` and `authorize` in #280. This allows for custom handling of launching the URL instead of only using `window.location.assign`. This allows for easier support of hybrid apps using things like Cordova or Capacitor.

**Note**: PR #273 sought to do this, but I re-worked it a bit based on the comments on my own fork. 

### References

- Issue: #273 
- Related PR: #280, #536 36

### Testing

`__tests__/index.test.ts` has been updated accordingly to mirror the changes made to `Auth0Client`.

- [x] This change adds test coverage for new/changed/fixed functionality

One notable thing was in the Jest tests, the `expectToHaveBeenCalledWithAuth0ClientParam` was catching two mocks (one from the prior function). The logged output below: 

```
      [
        [
          'https://test.auth0.com/v2/logout?query=params&auth0Client=eyJuYW1lIjoiYXV0aDAtc3BhLWpzIiwidmVyc2lvbiI6IjEuMTIuMSJ9&federated'
        ],
        [
          'https://test.auth0.com/v2/logout?query=params&auth0Client=eyJuYW1lIjoiX190ZXN0X2NsaWVudF9uYW1lX18iLCJ2ZXJzaW9uIjoiOS45LjkifQ%3D%3D'
        ]
      ]
```

This was bewildering to me and was only happening for this test: 

https://github.com/ironbow/auth0-spa-js/blob/6ff9ece3a78d5e7db855d37ef45e649184257a45/__tests__/index.test.ts#L2236-L2244
```javascript
    it('calls `window.location.assign` with the correct url with custom `options.auth0Client`', async () => {
      const auth0Client = { name: '__test_client_name__', version: '9.9.9' };
      const { auth0 } = await setup({ auth0Client });
      auth0.logout();
      expectToHaveBeenCalledWithAuth0ClientParam(
        window.location.assign,
        auth0Client
      );
    });
```
My means of getting around this was updating the helper function to references the last mock result in the array, rather than assuming there is only one:

https://github.com/ironbow/auth0-spa-js/blob/1054ba48d8054a97dda0ed6e6fbc83fd46b6b8f1/__tests__/helpers.ts#L2
```javascript
const url = (<jest.Mock>mock).mock.calls[mock.mock.calls.length - 1][0];
```
This may not be an appropriate fix, but I could not figure out what was causing it as the mocks should be cleared after each test.

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

### CODE REVIEW

File: __tests__/helpers.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 export const expectToHaveBeenCalledWithAuth0ClientParam = (mock, expected) => {
-  const [[url]] = (<jest.Mock>mock).mock.calls;
+  const url = (<jest.Mock>mock).mock.calls[mock.mock.calls.length - 1][0];
```
Comment (frederikprijck) [inline __tests__/helpers.ts:2]: The reason the original code was not working was because you have moved the `describe('logout()')` outside of the `describe('Auth0')`, meaning the mocks are not cleared and the test do not work when running in isolation. 

Nesting the describe blocks correctly again allows to revert this change while all tests work.

**Review by frederikprijck — CHANGES_REQUESTED** [review]

File: src/Auth0Client.ts [hunk 2]
```diff
@@ -695,6 +696,31 @@ export default class Auth0Client {
...
+  public buildLogoutUrl(options: LogoutUrlOptions = {}): string {
+    if (options.client_id !== null) {
+      options.client_id = options.client_id || this.options.client_id;
+    } else {
+      delete options.client_id;
+    }
+
+    const { federated, ...logoutOptions } = options;
+    this.cache.clear();
+    this.cookieStorage.remove('auth0.is.authenticated');
```
[the remark this answers is not in the document]
Comment (rnwolfe) [reply src/Auth0Client.ts:717]: Yes, you are correct. I misread the feedback in the other PR, but now see the correct approach. Will fix.

File: __tests__/helpers.ts (hunk 1 above)
Comment (rnwolfe) [reply __tests__/helpers.ts:2]: Thanks for catching this. I'm not sure how this occurred, but certainly the cause of the problem.

File: src/Auth0Client.ts [hunk 3]
```diff
@@ -695,6 +696,31 @@ export default class Auth0Client {
...
+   * Clears the application session and builds an `/v2/logout` URL for logout using
+   * the parameters provided as arguments.
+   * @param options
+   */
+  public buildLogoutUrl(options: LogoutUrlOptions = {}): string {
+    if (options.client_id !== null) {
+      options.client_id = options.client_id || this.options.client_id;
+    } else {
+      delete options.client_id;
+    }
```
[the remark this answers is not in the document]
Comment (rnwolfe) [reply src/Auth0Client.ts:713]: I don't think this would yield the desired results. As `buildLogoutUrl` can be called directly without using `logout`, if this is removed from `buildLogoutUrl` it would not always result in the expected `client_id` being added onto the logout URL parameters.

The only way around it would be to perform these checks in each function. Am I misinterpreting something?

File: src/Auth0Client.ts [hunk 4]
```diff
@@ -695,6 +696,31 @@ export default class Auth0Client {
...
+
+    const { federated, ...logoutOptions } = options;
```
[the remark this answers is not in the document]
Comment (rnwolfe) [reply src/Auth0Client.ts:715]: It shouldn't as I created a new type of `logoutUrlOptions` to not include `localOnly`. So, even if it is included as an option when `buildLogoutUrl` is called, it wouldn't end up as an available variable.

File: src/Auth0Client.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (frederikprijck) [reply src/Auth0Client.ts:713]: Let's keep this here but remove it from the logout method.

File: __tests__/index.test.ts [hunk 5]
```diff
@@ -2165,6 +2165,257 @@ describe('Auth0', () => {
     });
   });
 
+  describe('buildAuthorizeUrl', () => {
```
Comment (frederikprijck) [inline __tests__/index.test.ts:2168]: Why are there tests affected? I think we want to ensure we are only touching new tests or updating tests that needs to be updated.

**Review by frederikprijck — CHANGES_REQUESTED** [review]

**Comment by frederikprijck** [comment]
Added a few more comments. 

Thanks for the PR @rnwolfe , appreciate the work. I think we are almost to a point where we can get this approved.

File: __tests__/index.test.ts (hunk 5 above)
Comment (rnwolfe) [reply __tests__/index.test.ts:2168]: I'm not sure why these are being marked as changed.. whitespace changes? I'll try to resolve.

**Review by stevehobbsdev — CHANGES_REQUESTED** [review]

File: src/Auth0Client.ts [hunk 6]
```diff
@@ -695,6 +696,29 @@ export default class Auth0Client {
     return !!user;
   }
 
+  /**
+   * ```js
+   * await auth0.buildLogoutUrl(options);
+   * ```
+   *
+   * Clears the application session and builds an `/v2/logout` URL for logout using
```
[the remark this answers is not in the document]
Comment (rnwolfe) [reply src/Auth0Client.ts:704]: Yes, thanks. Good catch.

File: src/Auth0Client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rnwolfe) [reply src/Auth0Client.ts:715]: Yes, good points. Thanks for the reminder. 

I've simply deleted `options.localOnly` after it is no longer needed by `logout` but before sending the options to `buildLogoutUrl`. I think this should suffice. 

I made a quick attempt to make a test for this in jest, but I'm not that proficient in jest to properly mock/spy the `buildLogoutUrl` when calling `logout`. I would need a bit more to complete that. Do we think this is a necessary test?

File: __tests__/index.test.ts (hunk 5 above)
Comment (rnwolfe) [reply __tests__/index.test.ts:2168]: I think this happened in this commit: 8d2df480d6a245003e71b0584dc010baba294a17

I think it's just the result of how git aligned the changes when I erroneously messed up the test blocks. After fixing, it's still not lining up correctly. 

Not sure how to fix so the PR doesn't suggest the tests were changed? Any suggestions?

File: src/Auth0Client.ts [hunk 7]
```diff
@@ -731,11 +749,9 @@ export default class Auth0Client {
     if (localOnly) {
       return;
     }
-
-    const federatedQuery = federated ? `&federated` : '';
-    const url = this._url(`/v2/logout?${createQueryParams(logoutOptions)}`);
-
-    window.location.assign(`${url}${federatedQuery}`);
+    delete options.localOnly;
```
Comment (frederikprijck) [inline src/Auth0Client.ts:752]: We should not be doing this as this deletes the localOnly from the options object being passed to the logout, meaning any subsequent call to `logout` will now not have the localOnly property.

```
const options = {
  localOnly = true
};
client.logout(options); 
client.logout(options); // localOnly will not exist anymore
```

File: src/Auth0Client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (frederikprijck) [reply src/Auth0Client.ts:715]: We should not remove `options.localOnly`, see my other comment.

Why would you want to mock `buildLogoutUrl` ?

File: __tests__/index.test.ts (hunk 5 above)
Comment (frederikprijck) [reply __tests__/index.test.ts:2168]: This is not just a whitespace issue, the entire describe block for `buildAuthorizeUrl ` exists twice.

**Review by frederikprijck — CHANGES_REQUESTED** [review]

File: src/Auth0Client.ts (hunk 7 above)
Comment (rnwolfe) [reply src/Auth0Client.ts:752]: I didn't realize `delete` had this impact. 

Perhaps I'm confused on your desired outcome. Do you want to pass an object to `buildLogoutUrl` that contains or does not contain the `localOnly` option? 

If it should be passed, are you saying that we should just pull it out in the parsing of options so as to prevent it from being in the `logoutOptions` object, e.g.

```
const { federated, localOnly, ...logoutOptions } = options;
```

File: __tests__/index.test.ts (hunk 5 above)
Comment (rnwolfe) [reply __tests__/index.test.ts]: Thanks - fixed in 72253f6

File: src/Auth0Client.ts (hunk 7 above)
Comment (frederikprijck) [reply src/Auth0Client.ts:752]: My desired outcome is to not change any existing functionality by introducing this change to ensure we dont break anything for our users.

I think we should not pass a property `localOnly` to buildLogoutUrl as it has no meaning in that context.

I think we can use this:

```
const { localOnly, ...logoutOptions } = options;
if (localOnly && logoutOptions.federated) {
  ...
}
const url = this.buildLogoutUrl(logoutOptions);
```

File: src/Auth0Client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rnwolfe) [reply src/Auth0Client.ts:715]: Re: `localOnly` I've inquired about that on the other comment thread.

As far as the mock, if the desired outcome is if `logout({ localOnly: true})` is called that `localOnly` is not passed to `buildLogoutUrl`, I figured a test could be done to confirm it. Looking at the other tests, a mock was used with `.expectToBeCalledWith`. It was just a thought, though.

File: src/Auth0Client.ts (hunk 7 above)
Comment (rnwolfe) [reply src/Auth0Client.ts:752]: I see what you're saying. Thank you for the clarification.

File: __tests__/index.test.ts [hunk 8]
```diff
@@ -2281,106 +2323,106 @@ describe('Auth0', () => {
...
-    jest.spyOn(Auth0Client.prototype, 'getTokenSilently');
-    const getSpy = jest
-      .spyOn(require('../src/storage').CookieStorageWithLegacySameSite, 'get')
-      .mockReturnValueOnce(false);
+  describe('default creation function', () => {
+    it('does nothing if there is nothing in storage', async () => {
+      jest.spyOn(Auth0Client.prototype, 'getTokenSilently');
+      const getSpy = jest
+        .spyOn(require('../src/storage').CookieStorageWithLegacySameSite, 'get')
+        .mockReturnValueOnce(false);
```
Comment (frederikprijck) [inline __tests__/index.test.ts:2332]: The tests are starting to look better, can you please also remove any unnecessary whitespace and formatting changes?

**Review by frederikprijck — CHANGES_REQUESTED** [review]
A few more things before we can get this merged.
Thanks again for making all the changes!

File: src/Auth0Client.ts [hunk 9]
```diff
@@ -695,6 +696,29 @@ export default class Auth0Client {
     return !!user;
   }
 
+  /**
+   * ```js
+   * await auth0.buildLogoutUrl(options);
+   * ```
+   *
+   * Builds a `/v2/logout` URL for logout using the parameters provided as arguments.
+   * No session data is removed.
```
Comment (frederikprijck) [inline src/Auth0Client.ts:705]: I am not sure we should explicitly mention it is not removing session data. I prefer to mention what a method is doing, not what it is not doing

What do you think @stevehobbsdev ?

**Review by frederikprijck — CHANGES_REQUESTED** [review]

File: src/Auth0Client.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (frederikprijck) [reply src/Auth0Client.ts:715]: I do not think we have to worry about that in this PR. If you want, you can add a test but as we are not changing anything in this regard, I am fine with moving forward without a test for that.

Let me know if you want some help with adding a test, I don't mind giving you some guidance if you want to add a test for it.

[the remark this answers is not in the document]
Comment (rnwolfe) [reply src/Auth0Client.ts:715]: Thanks, I think the way you've suggested to handle it doesn't warrant a test (like you've said). 

I'll tackle that learning another time. 😃

File: src/Auth0Client.ts (hunk 9 above)
Comment (stevehobbsdev) [reply src/Auth0Client.ts]: Agreed. It's just building a URL, we don't really need to say much more than that. I think the method name already sets that expectation.

File: __tests__/index.test.ts (hunk 8 above)
Comment (rnwolfe) [reply __tests__/index.test.ts]: This should be fixed in 34964af

File: src/Auth0Client.ts (hunk 9 above)
Comment (rnwolfe) [reply src/Auth0Client.ts]: Fixed in a7197e0

File: __tests__/index.test.ts (hunk 8 above)
Comment (frederikprijck) [reply __tests__/index.test.ts]: Yes looks great!

**Comment by frederikprijck** [comment]
Thanks, this looks great @rnwolfe! 🎉  I will give this a last round of testing first thing tomorrow (getting a bit late here).

**Comment by rnwolfe** [comment]
Sounds good @frederikprijck! 👍 Thanks for your promptness. I'll keep an eye on things for tomorrow. 

I did just notice a failed cypress test in the CI on the last commit but it may be a fluke as it was successful in my environment.

**Comment by frederikprijck** [comment]
@rnwolfe CI is fine, it appears to have been a flake.

File: __tests__/index.test.ts [hunk 10]
```diff
@@ -2165,6 +2165,49 @@ describe('Auth0', () => {
...
+      });
+    });
+
+    it('creates correct query params when `options.federated` is true', async () => {
+      const { auth0, utils } = await setup();
+
+      auth0.buildLogoutUrl({ federated: true, client_id: null });
+      expect(utils.createQueryParams).toHaveBeenCalledWith({});
+    });
+  });
```
Comment (frederikprijck) [inline __tests__/index.test.ts:2209]: These tests all exist as part of the `logout` tests, I think we should remove those and keep the ones you added.

**Review by frederikprijck — CHANGES_REQUESTED** [review]
Looks good, one minor test cleanup and we should be ready to go.

File: src/Auth0Client.ts [hunk 11]
```diff
@@ -695,6 +696,28 @@ export default class Auth0Client {
     return !!user;
   }
 
+  /**
+   * ```js
+   * await auth0.buildLogoutUrl(options);
+   * ```
+   *
+   * Builds a `/v2/logout` URL for logout using the parameters provided as arguments.
```
Comment (stevehobbsdev) [inline src/Auth0Client.ts:704]: Let's tweak this a little:

```suggestion
   * Builds a URL to the logout endpoint using the parameters provided as arguments.
```

**Review by stevehobbsdev — CHANGES_REQUESTED** [review]

File: src/Auth0Client.ts (hunk 11 above)
Comment (rnwolfe) [reply src/Auth0Client.ts]: It's worth mentioning I kept this from the existing `logout` description and is similarly formatted `buildAuthorizeUrl` should that want to be updated one day.

**Comment by rnwolfe** [comment]
Made the last couple of changes 👍

**Review by frederikprijck — APPROVED** [review]

**Comment by frederikprijck** [comment]
@rnwolfe Looks good, could you merge in master ?

**Comment by rnwolfe** [comment]
@frederikprijck merged

**Review by stevehobbsdev — APPROVED** [review]
Thanks for the contribution! 🎉

**Comment by rnwolfe** [comment]
@frederikprijck sorry to bolt on to the PR, but when trying to install the new committed version in npm (`npm i --save https://github.com/auth0/auth0-spa-js.git#b27b3536486b23233d2b908721ac98f3632c14a2`), I get the following error:

```
[!] Error: Cannot find module 'rollup-plugin-node-resolve'
Require stack:
- /Users/rwolfe/.npm/_cacache/tmp/git-clone-49a30027/rollup.config.js
- /Users/rwolfe/.npm-global/lib/node_modules/rollup/dist/shared/loadConfigFile.js
- /Users/rwolfe/.npm-global/lib/node_modules/rollup/dist/bin/rollup
```

The same was true when I tried from my fork, but this seems off as I have it installed globally, as well as it being in the local `package.json`. The build works when I build manually in the cloned repo and then copy the dist into my `node_modules` which is obviously not great.

Am I missing something here? I'm trying to use this in my app prior to npm release. 

Thanks for any advice.

**Comment by frederikprijck** [comment]
I dont think this is supposed to be installed like that, as our artifacts published to npm are different from what we have in the repo (like we bundle things using rollup).

I think the easiest would be to checkout your branch locally (or the current master), build it and use npm or yarn link to link the `auth0-spa-js` to your local build output.

Something like this:

```
cd auth0-spa-js
npm run build
npm link
cd your-project
npm link auth0-spa-js
```

More info 
https://docs.npmjs.com/cli/link
https://medium.com/dailyjs/how-to-use-npm-link-7375b6219557

**Comment by rnwolfe** [comment]
@frederikprijck I see, thanks. I'll check that out for the time being.
